corehq/blobs/mixin.py

```diff
@@ -432,13 +443,25 @@ def deferred_put_attachment(self, content, name=None, content_type=None,
             "content_length": length,
         }
 
+    def deferred_delete_attachment(self, name):
+        """Mark attachment to be deleted on save"""
+        if self._deferred_blobs is None:
+            self._deferred_blobs = {}
+        self._deferred_blobs[name] = None
+
     def save(self):
         if self._deferred_blobs:
+            delete_names = []
             with self.atomic_blobs(super(DeferredBlobMixin, self).save):
                 # list deferred blobs to avoid modification during iteration
                 for name, info in list(self._deferred_blobs.iteritems()):
-                    self.put_attachment(name=name, **info)
-                assert not self._deferred_blobs, self._deferred_blobs
+                    if info is not None:
+                        self.put_attachment(name=name, **info)
+                    else:
+                        delete_names.append(name)
+            for name in delete_names:
+                self.delete_attachment(name)
```

Comment:
The code in the `atomic_blobs` context is executed [just before](https://github.com/dimagi/commcare-hq/blob/c334bd9ed7f1c8dd7a8ac8329237349a68d4b3a2/corehq/blobs/mixin.py#L206) the [doc is saved](https://github.com/dimagi/commcare-hq/blob/c334bd9ed7f1c8dd7a8ac8329237349a68d4b3a2/corehq/blobs/mixin.py#L207). This is outside the context so they are deleted after successful save. If we deleted them inside the context then they would be deleted even if the save failed.

It is possible that an exception could be raised just outside the context (after the doc is saved) before all deletes have gone through, which would result in orphaned blobs. I think that's not very likely, and better to err on the side of keeping data we don't need than deleting data we might need.